transport-classes-io_uring/src/main/java/io/netty/channel/uring/SubmissionQueue.java

```diff
@@ -0,0 +1,352 @@
+                throw new RuntimeException("SQ ring full and no submissions accepted");
+            }
+        }
+        long sqe = submissionQueueArrayAddress + (tail++ & ringMask) * SQE_SIZE;
+        return setData(sqe, op, flags, ioPrio, rwFlags, fd, bufferAddress, length, offset, udata);
+    }
+
+    private long setData(long sqe, byte op, int flags, short ioPrio, int rwFlags, int fd, long bufferAddress,
+                         int length, long offset, long udata) {
+        //set sqe(submission queue) properties
+
+        PlatformDependent.putByte(sqe + SQE_OP_CODE_FIELD, op);
+        PlatformDependent.putByte(sqe + SQE_FLAGS_FIELD, (byte) flags);
+        // This constant is set up-front
+        PlatformDependent.putShort(sqe + SQE_IOPRIO_FIELD, ioPrio);
+        PlatformDependent.putInt(sqe + SQE_FD_FIELD, fd);
+        PlatformDependent.putLong(sqe + SQE_OFFSET_FIELD, offset);
+        PlatformDependent.putLong(sqe + SQE_ADDRESS_FIELD, bufferAddress);
+        PlatformDependent.putInt(sqe + SQE_LEN_FIELD, length);
+        PlatformDependent.putInt(sqe + SQE_RW_FLAGS_FIELD, rwFlags);
+        PlatformDependent.putLong(sqe + SQE_USER_DATA_FIELD, udata);
+
+        if (logger.isTraceEnabled()) {
+            if (op == Native.IORING_OP_WRITEV || op == Native.IORING_OP_READV) {
+                logger.trace("add(ring {}): {}(fd={}, len={} ({} bytes), off={}, data={})",
+                        ringFd, Native.opToStr(op), fd, length, Iov.sumSize(bufferAddress, length), offset, udata);
+            } else {
+                logger.trace("add(ring {}): {}(fd={}, len={}, off={}, data={})",
+                        ringFd, Native.opToStr(op), fd, length, offset, udata);
+            }
+        }
+        return udata;
+    }
+
+    @Override
+    public String toString() {
+        StringJoiner sb = new StringJoiner(", ", "SubmissionQueue [", "]");
+        int pending = tail - head;
+        for (int i = 0; i < pending; i++) {
+            long sqe = submissionQueueArrayAddress + (head + i & ringMask) * SQE_SIZE;
+            sb.add(Native.opToStr(PlatformDependent.getByte(sqe + SQE_OP_CODE_FIELD)) +
+                    "(fd=" + PlatformDependent.getInt(sqe + SQE_FD_FIELD) + ')');
+        }
+        return sb.toString();
+    }
+
+    long addNop(int fd, int flags, short data) {
+        return enqueueSqe(Native.IORING_OP_NOP, flags, (short) 0, 0, fd, 0, 0, 0, data);
+    }
+
+    long addTimeout(int fd, long nanoSeconds, short extraData) {
+        setTimeout(nanoSeconds);
+        return enqueueSqe(Native.IORING_OP_TIMEOUT, 0, (short) 0, 0, fd,
+                timeoutMemoryAddress, 1, 0, extraData);
+    }
+
+    long addLinkTimeout(int fd, long nanoSeconds, short extraData) {
+        setTimeout(nanoSeconds);
+        return enqueueSqe(Native.IORING_OP_LINK_TIMEOUT, 0, (short) 0, 0, fd,
+                timeoutMemoryAddress, 1, 0, extraData);
+    }
+
+    long addPollIn(int fd) {
```

Comment:
We should cleanup this class and remove all the methods that are not used anymore. Usually people should use `IOUringIoRegistration` to submit stuff to the `SubmissionQueue`.

Reply:
done